src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
+
+    Promises variables of this type, ['a Lwt.t], are actually {b read-only} in
+    Lwt. Separate {e resolvers} of type ['a ]{!Lwt.u} are used to write to them.
+    Promises and their resolvers are created together by calling {!Lwt.task}.
+    There is one exception to this: most promises can be {e canceled} by calling
+    {!Lwt.cancel}, without going through a resolver. *)
+
+type 'a u
+(** Resolvers for promises of type ['a ]{!Lwt.t}.
+
+    Each resolver can be thought of as the {b write end} of one promise. It can
+    be passed to {!Lwt.wakeup_later}, {!Lwt.wakeup_later_exn}, or
+    {!Lwt.wakeup_later_result} to resolve that promise. *)
+
+val task : unit -> ('a t * 'a u)
+(** Creates a new pending {{: #TYPEt} promise}, paired with its {{: #TYPEu}
+    resolver}.
+
+    It is rare to use this function directly. Many helpers in Lwt, and Lwt-aware
+    libraries, call it internally, and return only the promise. You then chain
+    the promises together using {!Lwt.bind}.
+
+    However, it is important to understand [Lwt.task] as the fundamental promise
+    “constructor.” All other functions that evaluate to a promise can be, or
+    are, eventually implemented in terms of it. *)
+
+
+
+(** {3 Resolving} *)
+
+val wakeup_later : 'a u -> 'a -> unit
+(** [Lwt.wakeup_later r v] {e fulfills}, with value [v], the {e pending}
+    {{: #TYPEt} promise} associated with {{: #TYPEu} resolver} [r]. This
+    triggers callbacks attached to the promise.
+
+    If the promise is not pending, [Lwt.wakeup_later] raises
+    {{: https://caml.inria.fr/pub/docs/manual-ocaml/libref/Pervasives.html#VALinvalid_arg}
+    [Pervasives.Invalid_argument]}, unless the promise is {{: #VALcancel}
+    canceled}. *)
+
+val wakeup_later_exn : _ u -> exn -> unit
+(** [Lwt.wakeup_later_exn r exn] is like {!Lwt.wakeup_later}, except it
+    {e rejects} the associated {e pending} {{: #TYPEt} promise} with [exn]
+    instead. *)
 
 val return : 'a -> 'a t
-  (** [return e] is a thread whose return value is the value of the
-      expression [e]. *)
+(** [Lwt.return v] creates a new {{: #TYPEt} promise} that is {e already
+    fulfilled} with value [v].
+
+    This is needed to satisfy the type system in some cases. For example, in a
+    [match] expression where one case evaluates to a promise, the other cases
+    have to evaluate to promises as well:
+
+{[
+match need_input with
+| true -> Lwt_io.(read_line stdin)   (* Has type string Lwt.t... *)
+| false -> Lwt.return ""             (* ...so wrap empty string in a promise. *)
+]}
+
+    Another typical usage is in {{: #VALbind} [let%lwt]}. The expression after
+    the “[in]” has to evaluate to a promise. So, if you compute an ordinary
+    value instead, you have to wrap it:
+
+{[
+let%lwt line = Lwt_io.(read_line stdin) in
+Lwt.return (line ^ ".")
+]} *)
+
+val fail : exn -> _ t
+(** [Lwt.fail exn] is like {!Lwt.return}, except the new {{: #TYPEt} promise}
+    that is {e already rejected} with [exn]. *)
```

Comment:
The thing with `Lwt.fail` is that it's not actually as useful as `Lwt.return`.

1. `raise`d exceptions are captured in almost all contexts where Lwt is used, and converted into calls to `Lwt.fail` (or the equivalent) anyway.
2. `raise` trivially satisfies the type system.
3. The rest of the reference recommends `try%lwt`, and the major use of `Lwt.fail`, re-raising exceptions, is taken care of by that.

So I have a hard time introducing a non-misleading usage. That's why I find it easier to refer to `Lwt.return` without new examples :p

Reply:
I'm happy for people to educate me here, but I use `raise` a lot when programming with Lwt, and `fail` almost never. Also, I'd like to move away somewhat from exceptions, see earlier comment about favoring `Lwt_result`.

Reply:
I understand making the different comments about the different function somewhat proportionate to the importance that they have for the programmer. However, here the functions have similarities in terms of functionality/use (which is visible from the common part of their signatures).

When things share functionalities, explaining them with repetition is generally a good idea. Specifically, using *parallel syntax* helps the reader understand what's shared and what's not shared between the functions. See “parallel syntax” in Steven Pinker's The Sense of Style.

Maybe this type of explanations is better suited in a full API reference doc rather than an introductory tutorial. Is it intended more as the former of the latter?

Reply:
Point well taken.

> Maybe this type of explanations is better suited in a full API reference doc rather than an introductory tutorial. Is it intended more as the former of the latter?

Currently it's an incoherent mix of everything, but the specific part we are commenting on here, when it matures, is supposed to be an entry in a full API reference doc, whereas the introductory material will be moved out.